File: backend/models/health.py
"""Typed health-report models — the output of the deterministic rule engine.

Every field is rule-derived and reproducible: same input always produces the
same output. No LLM involvement at this stage.
"""
from __future__ import annotations

from pydantic import BaseModel, Field


class HealthScore(BaseModel):
    """Numeric health score with a human-readable label and display colour."""

    score: int = Field(ge=0, le=100)
    label: str
    color: str


class ProcessingLevel(BaseModel):
    """How processed the product is, determined by additive count and type."""

    id: str
    label: str
    description: str
    color: str


class PositiveIngredient(BaseModel):
    """An ingredient identified as beneficial."""

    name: str
    benefit: str


class ConcernItem(BaseModel):
    """An ingredient or additive flagged as a health concern."""

    name: str
    concern: str
    severity: str


class AllergenInfo(BaseModel):
    """A detected allergen and the ingredients that triggered it."""

    name: str
    triggered_by: list[str] = Field(default_factory=list)


class HealthConsideration(BaseModel):
    """A broader health consideration triggered by ingredient patterns."""

    title: str
    description: str
    type: str  # "warning" | "info"


class HealthReport(BaseModel):
    """Complete deterministic health report — the only shape this stage exposes."""

    health_score: HealthScore
    processing_level: ProcessingLevel
    positive_ingredients: list[PositiveIngredient] = Field(default_factory=list)
    ingredients_of_concern: list[ConcernItem] = Field(default_factory=list)
    health_considerations: list[HealthConsideration] = Field(default_factory=list)
    allergens: list[AllergenInfo] = Field(default_factory=list)
    unresolved_ingredients: list[str] = Field(default_factory=list)
    ingredient_count: int = 0
    resolved_count: int = 0

File: frontend/src/types/health.ts
/**
 * Health Report types — for future use when the backend adds
 * POST /api/v1/analyze and POST /api/v1/gemma endpoints.
 *
 * Currently the backend only exposes POST /ocr.
 * These types are ready for when the pipeline is complete.
 */

export interface HealthScore {
  score: number;
  label: string;
  color: string;
}

export interface ProcessingLevel {
  id?: string;
  label: string;
  description: string;
  color?: string;
}

export interface PositiveIngredient {
  name: string;
  benefit: string;
}

export interface ConcernItem {
  name: string;
  concern: string;
  severity: string;
}

export interface AllergenInfo {
  name: string;
  triggered_by: string[];
}

export interface HealthConsideration {
  title: string;
  description: string;
  type: string;
}

export interface UnresolvedFallbackItem {
  name: string;
  inferred_category: string;
  note: string;
  risk_indicator: 'neutral' | 'positive' | 'concern';
}

export interface RecommendedAlternative {
  name: string;
  category: string;
  reason: string;
  estimated_score?: number;
  url?: string;
}

export interface DeterministicReport {
  product_name?: string;
  health_score: HealthScore;
  processing_level: ProcessingLevel;
  positive_ingredients: PositiveIngredient[];
  ingredients_of_concern: ConcernItem[];
  health_considerations: HealthConsideration[];
  allergens: AllergenInfo[];
  recommended_alternatives?: RecommendedAlternative[];
  unresolved_ingredients: string[];
  unresolved_heuristics?: UnresolvedFallbackItem[];
  ingredient_count?: number;
  resolved_count?: number;
}

export interface GemmaMessage {
  role: 'user' | 'assistant';
  content: string;
}
